lib/Conversion/StandardToHandshake/InsertMergeBlocks.cpp

```diff
@@ -0,0 +1,376 @@
+    if (splitBlock == predDom)
+      // Needs no additional merge block, as this directly descends from the
+      // correct split block.
+      continue;
+
+    if (predsToConsider.count(splitBlock) == 0) {
+      // No other block with the same split block was found yet, so just store
+      // it and wait for a match.
+      predsToConsider.try_emplace(splitBlock, pred);
+      continue;
+    }
+
+    // Found a pair, so insert a new merge block for them.
+    Block *other = predsToConsider.lookup(splitBlock);
+    predsToConsider.erase(splitBlock);
+
+    Block *mergeBlock = buildMergeBlock(pred, other, currBlock, rewriter);
+    if (!mergeBlock)
+      return failure();
+
+    // Update info for the newly created block.
+    Block *splitIn = splitInfo.in.lookup(splitBlock);
+    splitInfo.in.try_emplace(mergeBlock, splitIn);
+    // By construction, this block has only one successor, therefore, out == in.
+    splitInfo.out.try_emplace(mergeBlock, splitIn);
+
+    preds.push_back(mergeBlock);
+  }
+  if (predsToConsider.size() != 0)
+    return currBlock->getParentOp()->emitError(
+        "irregular control flow is not yet supported");
+  return success();
+}
+
+/// Checks preconditions of this transformation.
+static LogicalResult preconditionCheck(Region &r,
+                                       ControlFlowLoopAnalysis &analysis) {
+  for (auto &info : analysis.topLevelLoops)
+    if (info.exitBlocks.size() > 1)
+      return r.getParentOp()->emitError(
+          "multiple exit nodes are not yet supported");
+
+  return success();
+}
+
+/// Insert additional blocks that serve as counterparts to the blocks that
+/// diverged the control flow.
+/// The resulting merge block tree is guaranteed to be a binary tree.
+///
+/// This transformation does not affect any blocks that are part of a loop as it
+/// treats a loop as one logical block.
+/// Irregular control flow is not supported and results in a failed
+/// transformation.
+LogicalResult circt::insertMergeBlocks(Region &r,
+                                       ConversionPatternRewriter &rewriter) {
+  Block *entry = &r.front();
+  DominanceInfo domInfo(r.getParentOp());
+
+  ControlFlowLoopAnalysis loopAnalysis(r);
+  if (failed(loopAnalysis.analyzeRegion()))
+    return failure();
+
+  if (failed(preconditionCheck(r, loopAnalysis)))
+    return failure();
+
+  // Traversing the graph in topological order can be simply done with a stack.
+  SmallVector<Block *> stack;
+  stack.push_back(entry);
+
+  // Holds the graph that contains the relevant blocks. It for example contracts
+  // loops into one block to preserve a DAG structure.
+  DualGraph graph(r, loopAnalysis);
+
+  // Counts the amount of predecessors remaining.
+  auto predsToVisit = graph.getPredCountMapCopy();
+
+  SplitInfo splitInfo;
+  splitInfo.in.try_emplace(entry, nullptr);
+
+  while (!stack.empty()) {
+    Block *currBlock = stack.pop_back_val();
+
+    Block *in;
+    Block *out;
```

Comment:
Suggestion: `nullptr` initialize + an assert later on, just for safety (although clang-tidy would complain if these are referenced before assignment).

Reply:
There is in fact a case where `in` isn't set, so initializing this to `nullptr` fixes a potential issue.